MetabolicAtlas/3d-network-viewer
# refactor: modernize code and move out helper functions

**Related issue(s) and PR(s)**  
This PR closes #35.



**Type of change**  

 
**List of changes made**  

- Rewrote `atlas-viewer-controls.js` as a `class`. Previously this was written as a function using things like `_this` and `prototype` to behave like a `class`. The official `three.js` has also made a similar change to their [TrackballControls.js](https://github.com/mrdoob/three.js/blob/master/examples/jsm/controls/TrackballControls.js), which our file is based on. Unlike the `TrackballControls.js` file that nests all of the code inside the `constructor`, the updated `atlas-viewer-controls.js` uses more standard practices such as class methods that hopefully makes the code more readable and easier to maintain. 
- Extracted parts from `met-atlas-viewer.js` that seems suitable to be placed into `helpers.js` The aim here was to find and move out pieces of code that do not contain key network manipulation logic, such as creating labels and materials. This should hopefully make it easier for developers to focus when working with `met-atlas-viewer.js`.
- Miscellaneous smaller changes such as replacing instances of `var` with `const/let` and using arrow functions where appropriate.


**Testing**  

1. Verify that the development runs: `npm start` and that the network loads as expected in the browser. Please also verify that there are no console errors.
2. Verify that the project can build without any problems: `npm run build`.

To test this out in the MetabolicAtlas repo, please refer to the following instructions.

1. Change value for dependency from version number to git branch in `frontend/package.json`.
```json
    "@metabolicatlas/3d-network-viewer": "MetabolicAtlas/3d-network-viewer#refactor/modernize-code",
```

2.  Install the dependency and build it.
```bash
ma-exec frontend yarn
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ install
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ run build
```

3. Use the code built from the previous step in `frontend/src/components/explorer/mapViewer/ThreeDviewer.vue`.
```javascript
// import { MetAtlasViewer } from '@metabolicatlas/3d-network-viewer';
import '@metabolicatlas/3d-network-viewer/public/met-atlas-viewer.js';

...

      // this.controller = MetAtlasViewer('viewer3d');
      this.controller = window.MetAtlasViewer.MetAtlasViewer('viewer3d');
```

**Further comments**  

- `flatShading`, which was used for creating `PointsMaterial` instances, is no longer a valid property so it has been removed.
- The `setConnectionsColor` function in `met-atlas-viewer.js` looks attractive for refactoring since it has four segments that look almost identical with the exception of some numbers. I'm a bit worried that making this less DRY would make it more difficult to understand what's going on. Please let me know if you think this should be refactored and I'm happy to add a commit for it for us to review.
- There is currently a convention where the a list of `@authors` is at the top of some files. I removed the list from `atlas-viewer-controls.js` since it has a mention of the official `three-js` file that it's based on. This is also not a practice that we commonly have elsewhere and personally I think it's unnecessary since there is a license on this repo and you can find all of the contributors for each file in the git history. I also find it a bit tedious to have to add your name to each file after modifying it. Please let me know if you would like to keep and update these or prefer to have them removed from all the files.
- There is an `atlas-viewer-controls.js.mod` file that I don't think is necessary anymore. Could we delete this from the project?
- Please let me know if you think more things should be refactored or if any of the new additions to `helpers.js` should be moved back to `met-atlas-viewer.js`.

**Definition of Done checklist**
